src/fs_watcher.rs

```diff
@@ -0,0 +1,75 @@
+use crate::DstSender;
+use crate::Endpoints;
+use inotify::{Event, EventMask, Inotify, WatchMask};
+use inotify_sys as ffi;
+use serde_json;
+use std::error::Error;
+use std::ffi::OsString;
+use std::mem;
+use std::string::String;
+use std::{fs, path};
+use tokio::stream::StreamExt;
+
+use crate::Dst;
+use crate::EndpointMeta;
+
+const EVENT_BUF_SZ: usize =
+    mem::size_of::<ffi::inotify_event>() + (libc::FILENAME_MAX as usize) + 1;
+
+#[derive(Debug)]
+pub struct FsWatcher {
+    endpoints_dir: String,
+    dst_sender: DstSender,
+}
+
+impl FsWatcher {
+    pub fn new(endpoints_dir: String, dst_sender: DstSender) -> Self {
+        Self {
+            endpoints_dir,
+            dst_sender,
+        }
+    }
+
+    #[tracing::instrument(skip(self), name = "FsWatcher::handle_event", level = "info")]
+    fn handle_event(
+        &mut self,
+        ev: Event<OsString>,
+    ) -> Result<(), Box<dyn Error + Send + Sync + 'static>> {
+        if let Some(file_name) = ev.name.and_then(|s| s.to_str().map(|s| s.to_string())) {
+            let dst: Dst = file_name.parse()?;
+            if ev.mask == EventMask::DELETE {
+                tracing::info!(?dst, "deleted");
+                self.dst_sender.delete_dst(dst);
+            } else {
+                let path = path::Path::new(&self.endpoints_dir).join(file_name);
```

Comment:
It might be worth taking `endpoints_dir` as a `PathBuf` rather than a `String` in the constructor, rather than borrowing it as a path here?